enhancements/kube-apiserver/kms-encryption-provider.md

```diff
@@ -0,0 +1,280 @@
+---
+title: kms-encryption-provider
+authors:
+  - "@swghosh"
+  - "@dgrisonnet"
+reviewers:
+  - "@TrilokGeer"
+  - "@tkashem"
+  - "@rvanderp"
+approvers:
+  - "@tkashem"
+  - "@deads2k"
+  - "@derekwaynecarr"
+api-approvers:
+  - "@JoelSpeed"
+creation-date: 2024-08-14
+last-updated: 2024-11-21
+status: implementable
+see-also:
+  - "/enhancements/kube-apiserver/encrypting-data-at-datastore-layer.md"
+  - "/enhancements/authentication/direct-external-oidc-provider.md"
+  - "/enhancements/installer/storage-class-encrypted.md"
+tracking-link:
+  - https://issues.redhat.com/browse/API-1684
+  - https://issues.redhat.com/browse/OCPSTRAT-108
+---
+
+
+# KMS Encryption Provider for etcd Secrets
+
+## Summary
+
+Provide a user-configurable interface to support encryption of data stored in etcd using a supported Key Management Service (KMS).
+
+## Motivation
+
+Today, we support local AES encryption at the datastore layer. It protects against etcd data leaks in the event of a etcd backup compromise. However, aescbc and aesgcm, which are supported encryption technologies today available in OpenShift do not protect against online host compromise i.e. in such cases, attackers can decrypt encrypted data from etcd using local keys, KMS managed keys protects against such scenarios.
+
+Users of OpenShift would like to encrypt secret data in etcd using self-managed KMS-backed keys.
+- https://issues.redhat.com/browse/OCPSTRAT-108
+
+### User Stories
+
+As an OpenShift administrator, I want to encrypt secrets in my cluster at rest using KMS keys so that I can comply with my organization's security requirements.
+
+As an OpenShift administrator, I want to let the KMS provider manage the lifecycle of the encryption keys. 
+
+As an OpenShift user, I enable encryption by setting apiserver.spec.encryption.type to kms.
+- After some time passes, a user makes a backup of etcd.
+- The user confirms that the secret values are encrypted by checking to see if they have the related KMS prefix.
+
+### Goals
+
+1. User can enable encryption (globally, with limited config)
+2. The keys used to encrypt are managed by the KMS instance (outside of the OpenShift control plane)
+3. The keys used to encrypt are periodically rotated automatically 
+4. Allow encryption to be disabled once it has been enabled
+5. Encryption at rest should not meaningfully degrade the performance of the cluster
+6. Allow the user to configure which KMS instance and keys are to be used, adopt KMSv2 from upstream
+
+### Non-Goals
+
+1. Allow the user to force key rotation
+2. Support for the complete lifecycle of KMS-managed keys directly within OpenShift control-plane
+3. Support for hardware security modules
+4. The user has an in-depth understanding of each phase of the encryption process
+5. Completely recover the cluster in the event of the KMS instance itself going down or keys getting lost
+6. Allow users to configure which resources will be encrypted
```

Comment:
Is this something planned for the future? 

I think this is a reasonable non-goal for a first pass, but taking a look at the `EncryptionConfiguration` resource type it appears to have support for this kind of configuration. I would imagine there is use case(s) for customers to have more control over the set of resources that are encrypted.

Reply:
No it is not right now and I don't remember seeing a RFE about that since we've supported encryption.
For OpenShift we are choosing which resources are encrypted and users just choose if they want to turn on encryption or not.
In case you've not seen the doc yet, https://docs.openshift.com/container-platform/4.17/security/encrypting-etcd.html shows the procedure to enable encryption and from a user perspective it is super simple and high-level.